Updates related to runfabtests.sh

shefty opened the pull request:
This series makes the following changes to using runfabtests:

1. Adds an exclusion file for the sockets provider.  Excludes dgram testing, since the udp providers has dgram ep support, and these tests regularly fail for the sockets provider.  Will expand exclusion list later to ignore tests that frequently fail cause CI failures when run over sockets.  The fabtests 'make test' option, used by travis CI, now references this exclusion file.

2. Changes the default provider used by runfabtests from sockets to tcp and udp.  Now if no provider is given, the tests will run over both of those providers, which will also pick up the rxm and rxd providers.

3. If no exclusions are given, runfabtests will 'search' for exclusions based on the provider name.  This only works if the test_configs directory is in the immediate path from the current working directory.  If no files can be found, run the test as normal.  This saves the user from typing '-f path_to_exclusion_file' in many cases.

4. Since all exclusion files are written using regular expressions, make that the default and remove it as an option.

5. Update the udp.exclude file to include ofi_rxd exclusions.

a-ilango commented on fabtests/scripts/runfabtests.sh:
```diff
@@ -596,15 +594,11 @@ function usage {
-	errcho -e " -R\tTreat test exclusions as regex patterns"
-	errcho -e "   \tNote: the test names / patterns for -e and -f options
-			would be matched with the list of test names defined
-			in this script. They don't have fi_ prefix"
```
We should probably keep this help text regarding the file patterns.

shefty replied:
I removed this since exclusions are always regular expressions now, so the fi_ prefix specified or not doesn't make a difference.

a-ilango replied:
Hmm, the user could also specify a stricter regex (e.g. ""^fi_cm_data"), that would only match tests that have correct prefix.

shefty replied:
I didn't realize that including the fi_ prefix didn't work.  I'll update the script to use the full binary name, which I think will be more intuitive.

a-ilango commented on fabtests/Makefile.am:
```diff
@@ -391,7 +391,7 @@ dist-hook: fabtests.spec
 
 test:
-	./scripts/runfabtests.sh -vvv -S $(os_excludes)
+	./scripts/runfabtests.sh -vvv -S $(os_excludes) -R -f ./test_configs/sockets/sockets.exclude sockets
```
I wonder if there's a way to get the exclude file from the location where it would be installed (e.g. /usr/share/fabtests/test_configs/)

shefty replied:
Maybe?  This part just copied the same syntax used by the other lines.  But I can update the script to also search /usr/share and /usr/local/share for exclude files.

a-ilango replied:
There's one potential problem though. If the installed config (in /usr/local) is from an older version of libfabric, the user might be unaware of that.

A better way of doing this would be to get the installation location from configure and use that. Not sure how easy is that. I think the present code that just searches the current working directory is good enough.

a-ilango commented on fabtests/scripts/runfabtests.sh:
```diff
@@ -595,9 +625,9 @@ case ${opt} in
 	g) GOOD_ADDR=${OPTARG}
 	;;
-	f) read_exclude_file ${OPTARG}
+	f) exclude_file=${OPTARG}
```
This would break specifying multiple exclude files. Makefile.am specifies some OS specific exclude file.

shefty replied:
Ah -- I'll need to figure out how to support multiple exclude files then.  My bash is the suck, so I missed that that feature was supported.

a-ilango replied:
The previous code where we call the function instead of setting a global exclude_file variable supports multiple exclude files.

shefty replied:
Yeah, I changed that because I wanted to reset the exclusions between running udp and tcp.  I'll redo this area so multiple files can be specified.

shefty commented:
PR updated with the following changes from v1:
1. Full test names now used by the test, including the fi_ prefix.
2. Multiple exclude files may be specified.  The -e and -f options are still exclusive, though we might be able to support that again if it's needed.

aingerson commented on fabtests/test_configs/udp/udp.exclude:
```diff
@@ -1,16 +1,19 @@
-mr_test
 cm_data
+
+# Issues in ofi_rxd provider prevent tests from passing
```
btw - rxd works fine with cntr_pingpong, cq_data, and poll
rdm_tagged_peek doesn't work, not because of issues in the provider per se but lacks support for it so it should probably go in the below section
IIRC, I included the tests that have polling/waiting in the exclude file because of Travis testing on OSX since it doesn't have epoll. Those should probably go in the osx.exclude file instead.

aingerson replied:
(and I should add that the cntr_pingpong test works fine *after* the PR I added fixing the rxd counters)

shefty replied:
Removed the cntr, cq_data, and poll tests from the exclude, moved tagged down.

shefty commented:
I updated the udp.exclude file to include a couple more unsupported tests.  I also reverted the change to the runfabtests.sh to completely remove the -R option.  That option is now simply ignored, which should keep external CI that uses -R from breaking.

a-ilango commented on fabtests/scripts/runfabtests.sh:
```diff
@@ -278,21 +281,54 @@ function read_exclude_file {
+	if [[ -n "$file_excludes" ]]; then
+		cur_excludes=${file_excludes}
+		return
+	fi
```
I think if we append the `file_excludes` to `cur_excludes` we can maintain the behavior of accepting both exclude file and exclude list on the command line.

shefty replied:
done
